torch/fx/experimental/symbolic_shapes.py

```diff
@@ -279,45 +334,69 @@ def magic_impl(self, other):
         out = func(expr, other_expr)
         out = sympy.expand(out)
         if method in ["truediv"]:
-            return PySymFloat(out, self.shape_env)
+            pytype = float
         else:
-            # TODO: relational operators actually technically return a
-            # PySymBool, this is a type error
-            return py_type(out, self.shape_env)
+            pytype = self.pytype
+
+        # TODO: relational operators actually technically return a
+        # PySymBool, this is a type error
+        return SymNode(out, self.shape_env, pytype)
 
     def unary_magic_impl(self):
         if SYM_FUNCTION_MODE:
             if method in ["ceil", "floor"]:
                 op = getattr(math, method)
             else:
                 op = getattr(operator, method)
-            return _handle_sym_dispatch(op, (self,), {})
+            r = _handle_sym_dispatch(op, (wrap_node(self),), {})
+            assert isinstance(r, (SymInt, SymFloat)), type(r)
+            return r.node
         # TODO: consider constant prop here
         expr = self.shape_env.replace(self.expr)
         out = func(expr)
         out = sympy.expand(out)
         if method in ["ceil", "floor"]:
-            return PySymInt(out, self.shape_env)
+            pytype = int
         else:
-            return py_type(out, self.shape_env)
+            pytype = self.pytype
+
+        return SymNode(out, self.shape_env, pytype)
 
-    # this should be wrapped transparently into torch.SymIntNode
     if method in unary_magic_methods:
-        setattr(py_type, method, unary_magic_impl)
-        setattr(py_type, f"__{method}__", unary_magic_impl)
+        setattr(SymNode, method, unary_magic_impl)
     else:
-        setattr(py_type, method, magic_impl)
-        setattr(py_type, f"__{method}__", magic_impl)
+        setattr(SymNode, method, binary_magic_impl)
+
+def _make_user_magic(method, user_type):
+    # User magic takes care of wrapping the other operand into a node,
+    # so that our internal logic can assume everything is nodes
+
+    def unary_magic_impl(self):
+        return wrap_node(getattr(self.node, method)())
+
+    def binary_magic_impl(self, other):
+        return wrap_node(getattr(self.node, method)(self.node.to_node(other)))
+
+    def rbinary_magic_impl(self, other):
+        return wrap_node(getattr(self.node.to_node(other), method)(self.node))
+
+    if method in unary_magic_methods:
+        setattr(user_type, f"__{method}__", unary_magic_impl)
+    else:
+        setattr(user_type, f"__{method}__", binary_magic_impl)
         if method in reflectable_magic_methods:
-            setattr(py_type, f"__r{method}__", magic_impl)
+            setattr(user_type, f"__r{method}__", rbinary_magic_impl)
+
+for method, func in magic_methods.items():
```

Comment:
nit: move this above _make_user_magic so that it is easier to see that they are independent.